lib/services/access_list.go

```diff
@@ -218,47 +233,165 @@ func UnmarshalAccessListMember(data []byte, opts ...MarshalOption) (*accesslist.
+func (a AccessListMembershipChecker) recursiveIsAccessListOwnerCheck(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) error {
+	seen := map[string]struct{}{}
+	type queueItem struct {
+		name  string
+		depth int
+	}
+	queue := []queueItem{{name: accessList.GetName(), depth: 0}}
+	ownershipErr := trace.NotFound("User '%s' is not an owner of Access List '%s', or a member in its nested list owners.", identity.Username, accessList.GetName())
+
+	// Ownership is inherited via Membership in nested Access Lists added as Owners,
+	// so we start by checking Owners of the top-level Access List for potential nested lists.
+	for _, owner := range accessList.GetOwners() {
+		if owner.MembershipKind == accesslist.MembershipKindList {
+			queue = append(queue, queueItem{name: owner.Name, depth: 1})
+		}
+	}
+
+	for len(queue) > 0 {
+		item := queue[0]
+		queue = queue[1:]
+
+		if item.depth > accesslist.MaxAllowedDepth {
+			return trace.LimitExceeded("Exceeded maximum depth of %d while checking for ownership in Access List '%s'.", accesslist.MaxAllowedDepth, accessList.GetName())
+		}
+
+		member, err := a.members.GetAccessListMember(ctx, item.name, identity.Username)
+		if err != nil && !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		if trace.IsNotFound(err) {
+			subAccessListMembers, err := a.GetAccessListNestedLists(ctx, item.name)
+			if err != nil {
+				return trace.NotFound("Error finding Access List '%s'.", item.name)
+			}
+			for _, next := range subAccessListMembers {
+				if _, ok := seen[next]; ok {
+					// Circular references should be skipped.
+					continue
+				}
+				seen[next] = struct{}{}
+				queue = append(queue, queueItem{name: next, depth: item.depth + 1})
+			}
+			continue
+		}
+
+		expires := member.Spec.Expires
+		if !expires.IsZero() && !a.clock.Now().Before(expires) {
+			// Avoid non-deterministic behavior here: If user's membership is expired, then
+			// continue checking, in case their membership in a related list is still valid.
+			ownershipErr = trace.AccessDenied("User '%s's membership has expired in Access List '%s'.", identity.Username, item.name)
+			continue
+		}
+
+		subAccessList, err := a.members.GetAccessList(ctx, item.name)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if UserMeetsRequirements(identity, subAccessList.Spec.MembershipRequires) {
+			return nil
+		}
+	}
+
+	return ownershipErr
+}
+
+// IsAccessListMember will determine whether the provided tlsca.Identity is a member for the accessList
+// and meets its membership requirements.
+//
+// If membership is inherited (through the tlsca.Identity having Membership in a nested Access List),
+// the tlsca.Identity must also meet the membership requirements of the nested list.
+func (a AccessListMembershipChecker) IsAccessListMember(ctx context.Context, identity tlsca.Identity, accessList *accesslist.AccessList) (MembershipOrOwnershipType, error) {
```

Comment:
Same question as above - the Hierarchy already implements this logic, can we reuse?